Tests/Realm.Tests/Database/AsyncTests.cs

```diff
@@ -155,5 +158,405 @@ static async Task<long> MeasureTiming(Func<Task> func)
+                Assert.That(queryable.Count, Is.EqualTo(1));
+                Assert.That(queryable.First().FullName, Is.EqualTo(personName));
+            });
+        }
+
+        [Test]
+        public void AsyncBeginWrite_StackReferencedNativeException_Does_Not_Throw()
+        {
+            TestHelpers.RunAsyncTest(async () =>
+            {
+                using var tokenSource2 = new CancellationTokenSource();
+                CancellationToken ct = tokenSource2.Token;
+                var t = Task.Run(async () =>
+                {
+                    var realm = GetRealm();
+                    var transaction = await realm.BeginWriteAsync();
+                    while (true)
+                    {
+                        await Task.Delay(1000);
+                    }
+                }, ct);
+
+                // give time for the lock to be acquired from the other task
+                await Task.Delay(2000);
+
+                Exception ex = null;
+
+                try
+                {
+                    await _realm.BeginWriteAsync();
+                }
+                catch (Exception e)
+                {
+                    ex = e;
+                }
+
+                Assert.That(ex, Is.Null);
+                tokenSource2.Cancel();
+            });
+        }
+
+        [Test]
+        public void AsyncBeginWrite_Throws_User_Exception()
+        {
+            TestHelpers.RunAsyncTest(async () =>
+            {
+                InvalidCastException ex = null;
+                try
+                {
+                    await _realm.WriteAsync(() =>
+                    {
+                        char ch = Convert.ToChar(true);
+                    });
+                }
+                catch (InvalidCastException e)
+                {
+                    ex = e;
+                }
+
+                Assert.That(ex, Is.InstanceOf<InvalidCastException>().And.Message.EqualTo("Invalid cast from 'Boolean' to 'Char'."));
+            });
+        }
+
+        [Test]
+        public void AsyncWrite_Fifo_Order_Respected()
+        {
+            TestHelpers.RunAsyncTest(async () =>
+            {
+                Person[] people = new[]
+                {
+                    new Person { FirstName = "Anderson" },
+                    new Person { FirstName = "Banderson" },
+                    new Person { FirstName = "Canderson" },
+                    new Person { FirstName = "Danderson" }
+                };
+
+                ConcurrentQueue<Person> markers = new();
+                Queue<Person> actualWriters = new();
+                Task[] tasks = new Task[people.Length];
+                _realm.Write(() =>
+                {
+                    _realm.RemoveAll<Person>();
+                });
+
+                var thread = new AsyncContextThread();
+                Parallel.For(0, people.Length, index =>
+                {
+                    tasks[index] = thread.Factory.Run(async () =>
```

Comment:
You're scheduling work on the same thread, so I don't think this test is testing our internal FIFO guarantees.

Reply:
The scope doc says
> Ensure FIFO ordering of async write transactions originating from the same Scheduler

And talking with Jørgen, he also confirmed that the async write/commit is supposed to all happen within the same thread, which for us translates in "the same synchronization context". So I think that the test is correct.
If you think I'm wrong, you could explain in a call, as by writing I think it'd take too long.

Reply:
Hm... in this case the test can be a lot simpler then - I got thrown off by the `Parallel.For` here. A simpler test would be:

```csharp
var acquisitionOrder = new List<int>();
var tasks = new List<Task>();

for (var i = 0; i < 5; i++)
{
    var index = i;
    tasks.Add(realm.WriteAsync(() =>
    {
        acquisitionOrder.Add(index);
    });
}

await Task.WhenAll(tasks);

Assert.That(acquisitionOrder, Is.EqualTo(new[] { 0, 1, 2, 3, 4 }));
```